packages/keystone/healthCheck.js

```diff
@@ -0,0 +1,227 @@
+/**
+ * This service defines custom api handle which will output health status of defined services
+ *
+ *
+ * ## Simple usage:
+ *
+ * Go to <url>
+ *
+ * Example: if all checks pass:
+ * 200
+ * {
+ *     postgres: pass
+ *     redis: pass
+ *     apollo: pass
+ * }
+ *
+ * Example: if any check fails
+ * 500
+ * {
+ *     postgres: pass
+ *     redis: pass
+ *     apollo: fail
+ * }
+ *
+ *
+ * ## Advanced usage:
+ *
+ * API supports to filter checks you need via query string
+ *
+ * Syntax: <url>?checks=<check1>,<check2>,...,<checkN>
+ *
+ * Configuration example - available checks: postgres, redis, apollo, postgres pass, apollo fails.
+ *
+ * Request: <url>?checks=postgres
+ * 200
+ * {
+ *      postgres: pass
+ * }
+ *
+ * Request <url>?checks=postgres,redis
+ * 500
+ * {
+ *      postgres: pass,
+ *      redis: fail
+ * }
+ *
+ * Note: if configured, non specified checks are ignored. That means, if redis and apollo fails, output will be 200OK, since only postgres was specified
+ *
+ * ## Response codes:
+ *
+ * API will output 500 if ANY check fails
+ * API will output 200 if EVERY check passes
+ * API will output 400 if checks=... is badly configured
+ *
+ *
+ * ## Create own checks:
+ *
+ * Check is JS object with some required keys:
+ *
+ * {
+ *     *name: "postgres"                        // name of the check
+ *     *run: () => {bool}                       // check function. should output boolean
+ *     prepare: async ({ keystone }) => {...}   // prepare function. is called during prepareMiddleware() stage
+ * }
+ *
+ */
+
+const express = require('express')
+const LRUCache = require('lru-cache')
+
+const { getRedisClient } = require('./redis')
+
+const DEFAULT_HEALTHCHECK_URL = '/server-health'
+const DEFAULT_INTERVAL = 1000 // 1s
+const DEFAULT_TIMEOUT = 5000 // 5s
+const DELIMITER = ','
+const PASS = 'pass'
+const FAIL = 'fail'
+
+const BAD_REQUEST = 400
+const OK = 200
+const INTERNAL_SERVER_ERROR = 500
+
+const getRedisHealthCheck = (clientName = 'healthcheck') => {
+    return {
+        name: 'redis',
+        prepare: () => {
+            this.redisClient = getRedisClient(clientName)
+        },
+        run: async () => {
+            try {
+                const client = this.redisClient
+                const res = await client.ping()
+                if (res === 'PONG') { return true }
+            } catch (e) { return false }
```

Comment:
making check function return optional error (success is when there is no error ) may make this thing move informative?